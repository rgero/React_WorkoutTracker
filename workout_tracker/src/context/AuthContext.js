import AsyncStorage from '@react-native-async-storage/async-storage';

import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const authReducer = (state, action) => {
    switch (action.type)
    {
        case "addError":
            return { ...state, errorMessage: action.payload};
        case "storeToken":
            return { errorMessage: "", token: action.payload};
        case "clearErrorMessage":
            return {...state, errorMessage: ""}
        case "signout":
            return { token: null, errorMessage: ""}
        default:
            return state;
    }
};

const tryLocalSignin = dispatch => async () => {
    const token = await AsyncStorage.getItem('token');
    if (token)
    {
        dispatch({ type: 'storeToken', payload: token})
    }
}

const clearErrorMessage = dispatch => () => {
    dispatch( {type: 'clearErrorMessage' })
}

const signUp = dispatch => {
    return async ({ email, displayName, password }) => {
        try {
            const response = await trackerAPI.post('/signup', { email, displayName, password });
            
            // Storing the Token
            await AsyncStorage.setItem('token', response.data.token);
            dispatch({type: "storeToken", payload: response.data.token})
        } catch (err) {
            console.log(err.message);
            dispatch({ type: 'addError', payload: "Sign-up failed. Please try again"})
        }
    };
}

const signIn = (dispatch) => {
    return async ({ email, password }) => {
        try {
            const response = await trackerAPI.post('/signin', { email, password });
            
            // Storing the Token
            await AsyncStorage.setItem('token', response.data.token);
            dispatch({type: "storeToken", payload: response.data.token})
        } catch (err) {
            console.log(err.message);
            dispatch({ type: 'addError', payload: "Login failed. Please try again"})
        }
    };
}

const signOut = dispatch => async () => {
    try {
        await AsyncStorage.removeItem('token');
        dispatch({type: "signout"})
    } catch (err)
    {
        console.log(err);
    }
}

export const {Provider, Context } = createDataContext(authReducer, 
                                                      {clearErrorMessage, signUp, signIn, signOut, tryLocalSignin}, 
                                                      {
                                                        token: null,
                                                        errorMessage: ""
                                                      }
                                                    )