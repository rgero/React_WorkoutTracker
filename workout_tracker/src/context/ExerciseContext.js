import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const exerciseReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        default:
            return state;
    }
};

const fetchExercises = dispatch => async () => {
    const response = await trackerAPI.get('/exercises');
    dispatch({type: 'fetch', payload: response.data})
};

const fetchExerciseByID = dispatch => async (targetID) => {
    const response = await trackerAPI.get(`/exercises/${targetID}`);
    dispatch({type: 'fetch', payload: response.data})
};


export const {Provider, Context} = createDataContext(
    exerciseReducer,
    {fetchExercises, fetchExerciseByID},
    []
);