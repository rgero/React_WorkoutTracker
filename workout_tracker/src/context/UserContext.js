import AsyncStorage from '@react-native-async-storage/async-storage';

import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const userReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        default:
            return state;
    }
};

const getUser = dispatch => async () => {
    const token = await AsyncStorage.getItem('token');
    if (token)
    {
        const response = await trackerAPI.get('/users');
        dispatch({type: 'fetch', payload: response.data})
    }
}

export const {Provider, Context} = createDataContext(
    userReducer,
    {getUser},
    []
);