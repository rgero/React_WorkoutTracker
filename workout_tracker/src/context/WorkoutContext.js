import trackerAPI from '../api/workouttracker';
import createDataContext from "./createDataContext";

const workoutReducer = (state, action) => {
    switch(action.type)
    {
        case 'fetch':
            return action.payload;
        case "create":
            return state.concat(action.payload);
        case "edit":
            return state.map( (workout) => {
                if (workout._id === action._id) {
                    return {
                        ...workout,
                        ...action.update
                    }
                } else {
                    return workout;
                }
            })
        case "delete":
            return state.filter( (testWorkout) => {
                return action.id !== testWorkout._id;
            });
        default:
            return state;
    }
};

const fetchWorkouts = dispatch => async () => {
    const response = await trackerAPI.get('/workouts');
    dispatch({type: 'fetch', payload: response.data})
};

const createWorkout = dispatch => async (workout) => 
{
    const response = await trackerAPI.post('/workouts', workout)
    dispatch({type: "create", payload: response.data});
};

const updateWorkout = dispatch => async (workout) =>
{
    const response = await trackerAPI.put(`/workouts/${workout._id}`, workout);
    dispatch({type: "edit", payload: response.data});
}

const deleteWorkout = dispatch => async (targetID) => 
{
    const response = await trackerAPI.delete(`/workouts/${targetID}`)
    dispatch({type: "delete", id: targetID});
}

export const {Provider, Context} = createDataContext(
    workoutReducer,
    {fetchWorkouts, createWorkout, deleteWorkout, updateWorkout},
    []
);