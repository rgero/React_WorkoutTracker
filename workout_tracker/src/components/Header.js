import React, { useContext, useEffect } from 'react';
import Container from 'react-bootstrap/Container';
import Nav from 'react-bootstrap/Nav';
import Navbar from 'react-bootstrap/Navbar';

import {Context as AuthContext} from '../context/AuthContext';
import {Context as UserContext} from '../context/UserContext';

function NavigationBar() {
    const {state, tryLocalSignin, signOut} = useContext(AuthContext);
    const {state: user, getUser} = useContext(UserContext);

    useEffect(()=> {
        const processUser = async () => {
            tryLocalSignin();
            await getUser();
        }
        processUser();
    }, [])
    return (
        <Navbar bg="dark" expand="lg" variant="dark">
            <Container>
                <Navbar.Brand href="/">Workout Tracker</Navbar.Brand>
                <Navbar.Toggle aria-controls="basic-navbar-nav" />
                <Navbar.Collapse id="basic-navbar-nav">
                    { state.token ? (
                        <>
                            
                            <Nav className="me-auto">
                                <Nav.Link href="/dashboard">Dashboard</Nav.Link>
                                <Nav.Link href="/create">Add Workout</Nav.Link>
                            </Nav>
                            <Nav>
                                { user ? (
                                    <Nav.Link>
                                        {user.displayName}
                                    </Nav.Link>
                                ) : ( 
                                    null 
                                )}
                                <Nav.Link onClick={signOut}>Sign Out</Nav.Link>
                            </Nav>
                        </>
                    ) : (
                        <Nav className="ms-auto">
                            <Nav.Link href="/login">Login</Nav.Link>
                            <Nav.Link href="/signup">Sign Up</Nav.Link>
                        </Nav>
                    )}
                </Navbar.Collapse>
            </Container>
        </Navbar>
    );
}

export default NavigationBar;