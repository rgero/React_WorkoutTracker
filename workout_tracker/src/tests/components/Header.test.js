import TestRenderer from 'react-test-renderer';
import Header from '../../components/Header'
import {Context as AuthContext} from '../../context/AuthContext';


describe("No User Logged In", ()=> {
    let element;
    beforeAll(()=> {
        const testValue = {
            state: {token: null},
            tryLocalSignin: jest.fn(),
            signOut: jest.fn()
        }
        element = new TestRenderer.create(
            <AuthContext.Provider value={testValue}>
                <Header />
            </AuthContext.Provider>
        );
    });

    test("Matches Snapshot", () => {  
        expect(element.toJSON()).toMatchSnapshot();
    });

    test("Correct button count", () => {
        let buttons = element.root.findAllByType("a");
        expect(buttons.length).toEqual(3);
    });
})

describe("User Logged In, with fake token", ()=> {
    let element;
    beforeAll(()=> {
        const testValue = {
            state: {token: "FakeTokenIsNeat"},
            tryLocalSignin: jest.fn(),
            signOut: jest.fn()
        }
        element = new TestRenderer.create(
            <AuthContext.Provider value={testValue}>
                <Header />
            </AuthContext.Provider>
        );
    });


    test("Matches Snapshot", () => {  
        expect(element.toJSON()).toMatchSnapshot()
    })

    test("Matches button count", ()=> {
        let buttons = element.root.findAllByType("a");
        expect(buttons.length).toEqual(4);
    })
})

test('Branding button works', ()=> {
    const testValue = {
        state: {token: null},
        tryLocalSignin: jest.fn(),
        signOut: jest.fn()
    }
    const element = new TestRenderer.create(
        <AuthContext.Provider value={testValue}>
            <Header />
        </AuthContext.Provider>
    );

    let buttons = element.root.findAllByType("a");
    let headerButton = buttons[0].props;
    let expectedHeader = {
        href: '/',
        children: 'Workout Tracker',
        className: 'navbar-brand'
    }
    expect(headerButton).toEqual(expectedHeader);
})

