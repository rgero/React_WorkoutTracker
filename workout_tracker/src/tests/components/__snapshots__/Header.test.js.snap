// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`No User Logged In Matches Snapshot 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <div
    className="container"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Workout Tracker
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="ms-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rr-ui-event-key="/login"
          disabled={false}
          href="/login"
          onClick={[Function]}
        >
          Login
        </a>
        <a
          className="nav-link"
          data-rr-ui-event-key="/signup"
          disabled={false}
          href="/signup"
          onClick={[Function]}
        >
          Sign Up
        </a>
      </div>
    </div>
  </div>
</nav>
`;

exports[`User Logged In, with fake token Matches Snapshot 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <div
    className="container"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Workout Tracker
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="me-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rr-ui-event-key="/dashboard"
          disabled={false}
          href="/dashboard"
          onClick={[Function]}
        >
          Dashboard
        </a>
        <a
          className="nav-link"
          data-rr-ui-event-key="/create"
          disabled={false}
          href="/create"
          onClick={[Function]}
        >
          Add Workout
        </a>
      </div>
      <div
        className="navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="nav-link"
          data-rr-ui-event-key="/user"
          disabled={false}
          href="/user"
          onClick={[Function]}
        >
          Roy
        </a>
        <a
          className="nav-link"
          data-rr-ui-event-key={null}
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Sign Out
        </a>
      </div>
    </div>
  </div>
</nav>
`;
